activerecord/lib/active_record/connection_handling.rb

```
@@ -47,6 +47,95 @@ module ConnectionHandling
+    #
+    #   class AnimalsModel < ApplicationRecord
+    #     self.abstract_class = true
+    #
+    #     connects_to database: { writing: :primary, reading: :primary_replica }
+    #   end
+    #
+    # Returns an array of established connections.
+    def connects_to(database: {})
+      connections = []
+
+      database.each do |handler_key, database_key|
+        config_hash = resolve_config_for_connection(database_key)
+        handler = lookup_connection_handler(handler_key)
+
+        connections << handler.establish_connection(config_hash)
+      end
+
+      connections
+    end
+
+    # Connects to a database or handler for the duration of the block.
+    #
+    # If a handler is passed, Active Record will look up the connection
+    # based on the requested handler:
+    #
+    #   ActiveRecord::Base.connected_to(handler: :writing) do
+    #     Dog.create! # creates dog using dog connection
+    #   end
+    #
+    #   ActiveRecord::Base.connected_to(handler: :reading) do
+    #     Dog.create! # throws exception because we're on a replica
+    #   end
+    #
+    #   ActiveRecord::Base.connected_to(handler: :unknown_handler) do
+    #     # raises exception due to non-existent handler
+    #   end
+    #
+    # For cases where you may want to connect to a database outside of the model,
+    # you can use +connected_to+ with a +database+ argument. The +database+ argument
+    # expects a hash consisting of a +handler_key+ and a +database_key+.
+    #
+    # This will connect to a new database for the queries inside the block.
+    #
+    #   ActiveRecord::Base.connected_to(database: { slow_replica: :animals_slow_replica }) do
+    #     Dog.run_a_long_query # runs a long query while connected to the +animals_slow_replica+
+    #   end
+    def connected_to(database: nil, handler: nil, &blk)
+      if database && handler
+        raise ArgumentError, "connected_to can only accept handler or database, but not both arguments."
+      end
```

Should we raise an exception if both are `nil`?